include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,276 @@
+        auto offset = r.offset_;
+        timing_t new_value = t;
+        if (time_by_offset_.count(offset) == 0) {
+          // ignore the 1st timing to cover for JIT compilation
+          time_by_offset_[offset] = time_data_t{0, std::numeric_limits<timing_t>::max()};
+        } else {
+          auto &td = time_by_offset_[offset];
+          auto n = td.num_timings_;
+          new_value = (n*td.value_+t)/(n+1);
+          td.num_timings_ = n+1;
+          td.value_ = new_value;
+        }
+        if (new_value < best_timing_) {
+          best_timing_ = new_value;
+          best_resource_ = r;
+        }
+      }
+
+    };
+
+    template <typename Resource, typename Tuner>
+    class auto_tune_selection_type {
+      using policy_t = auto_tune_policy<Backend, KeyArgs...>;
+      policy_t policy_;
+      using resource_with_offset_t = Resource;
+      resource_with_offset_t resource_;
+      using tuner_t = Tuner;
+      std::shared_ptr<tuner_t> tuner_;
+
+    public:
+      auto_tune_selection_type() : policy_(deferred_initialization) {}
+
+      auto_tune_selection_type(const policy_t& p, resource_with_offset_t r, std::shared_ptr<tuner_t> t)
+        : policy_(p), resource_(r), tuner_(t) {}
+
+      auto unwrap() { return ::oneapi::dpl::experimental::unwrap(resource_.r_); }
+
+      policy_t get_policy() { return policy_; };
+
+      void report(const execution_info::task_time_t&, const typename execution_info::task_time_t::value_type& v) const {
+        tuner_->add_new_timing(resource_, v);
+      }
+    };
+
+   public:
+
+    // Needed by Policy Traits
+    using resource_type = decltype(unwrap(std::declval<wrapped_resource_t>()));
+    using wait_type = typename Backend::wait_type;
+    using selection_type = auto_tune_selection_type<resource_with_offset_t, tuner_t>;
+
+    auto_tune_policy(double resample_time=never_resample) {
+      if (resample_time != deferred_initialization) {
+        initialize(resample_time);
+      }
+    }
+
+    auto_tune_policy(const std::vector<resource_type>& u, double resample_time=never_resample) {
+      if (resample_time != deferred_initialization) {
+        initialize(u, resample_time);
+      }
+    }
+
+    void initialize(double resample_time=never_resample) {
+      if (!state_) {
+        state_ = std::make_shared<state_t>();
+        backend_ = std::make_shared<Backend>();
+        initialize_impl(resample_time);
+      }
+    }
+
+    void initialize(const std::vector<resource_type>& u, double resample_time=never_resample) {
+      if (!state_) {
+        state_ = std::make_shared<state_t>();
+        backend_ = std::make_shared<Backend>(u);
+        initialize_impl(resample_time);
+      }
+    }
+
+    template<typename Function, typename ...Args>
+    selection_type select(Function&& f, Args&&...args) {
+      if (state_) {
+        std::unique_lock<std::mutex> l(state_->m_);
+        auto k =  make_task_key(std::forward<Function>(f), std::forward<Args>(args)...);
```

Comment:
Looks like double search on ```tuner_by_key_``` inside the function ```make_task_key``` and on next line.

Reply:
Not changing for now